lib/facter/util/consul.rb

```diff
@@ -0,0 +1,80 @@
+module Facter::Util::Consul
+  require 'json'
+  require 'net/http'
+  require 'uri'
+
+  CONSUL_URL = 'http://localhost:8500/v1'
```

Comment:
The big downside to this is that it makes a lot of assumptions about the host/port/auth/etc that the rest of this puppet module does not assume.

I'm also very sensitive to facts doing lots of work on servers on every puppet run unnecessarily. 

Can you tell me more about your use-case of this fact?

Reply:
It's a port of our internal fact that we rely on a lot for discovering databases etc in multi-tier applications (with masterless puppet) and for leader election of clustered services that do not have their own quorum protocol. For example:
```
# facter -p consul_services
{
  consul => {
    nodes => [
      "10.129.4.39:8300",
      "10.129.5.194:8300",
      "10.129.7.210:8300"
    ],
    nodes_string => "10.129.4.39:8300,10.129.5.194:8300,10.129.7.210:8300",
    is_current_node_leader => false
  },
  elasticsearch_exporter => {
    nodes => [
      "10.129.4.39:9108",
      "10.129.5.106:9108",
      "10.129.6.110:9108",
      "10.129.6.232:9108",
      "10.129.7.233:9108",
      "10.129.5.194:9108",
      "10.129.7.210:9108",
      "10.129.4.85:9108",
      "10.129.7.92:9108",
      "10.129.5.235:9108",
      "10.129.4.135:9108"
    ],
    nodes_string => "10.129.4.39:9108,10.129.5.106:9108,10.129.6.110:9108,10.129.6.232:9108,10.129.7.233:9108,10.129.5.194:9108,10.129.7.210:9108,10.129.4.85:9108,10.129.7.92:9108,10.129.5.235:9108,10.129.4.135:9108",
    is_current_node_leader => false
  },
  graylog => {
    nodes => [
      "10.129.5.106:9000",
      "10.129.6.110:9000",
      "10.129.6.232:9000",
      "10.129.7.233:9000",
      "10.129.4.85:9000",
      "10.129.7.92:9000",
      "10.129.5.235:9000",
      "10.129.4.135:9000"
    ],
    nodes_string => "10.129.5.106:9000,10.129.6.110:9000,10.129.6.232:9000,10.129.7.233:9000,10.129.4.85:9000,10.129.7.92:9000,10.129.5.235:9000,10.129.4.135:9000",
    is_current_node_leader => true
  },
  mongodb => {
    nodes => [
      "10.129.6.96:27017"
    ],
    nodes_string => "10.129.6.96:27017",
    is_current_node_leader => false
  },
  mongodb_exporter => {
    nodes => [
      "10.129.6.96:9216"
    ],
    nodes_string => "10.129.6.96:9216",
    is_current_node_leader => false
  },
  node_exporter => {
    nodes => [
      "10.129.4.39:9100",
      "10.129.5.106:9100",
      "10.129.6.110:9100",
      "10.129.6.96:9100",
      "10.129.6.232:9100",
      "10.129.7.233:9100",
      "10.129.5.194:9100",
      "10.129.7.210:9100",
      "10.129.4.85:9100",
      "10.129.7.92:9100",
      "10.129.5.235:9100",
      "10.129.4.135:9100"
    ],
    nodes_string => "10.129.4.39:9100,10.129.5.106:9100,10.129.6.110:9100,10.129.6.96:9100,10.129.6.232:9100,10.129.7.233:9100,10.129.5.194:9100,10.129.7.210:9100,10.129.4.85:9100,10.129.7.92:9100,10.129.5.235:9100,10.129.4.135:9100",
    is_current_node_leader => false
  },
  prometheus => {
    nodes => [
      "10.129.5.106:9090",
      "10.129.6.110:9090",
      "10.129.6.232:9090",
      "10.129.7.233:9090",
      "10.129.4.85:9090",
      "10.129.7.92:9090",
      "10.129.5.235:9090",
      "10.129.4.135:9090"
    ],
    nodes_string => "10.129.5.106:9090,10.129.6.110:9090,10.129.6.232:9090,10.129.7.233:9090,10.129.4.85:9090,10.129.7.92:9090,10.129.5.235:9090,10.129.4.135:9090",
    is_current_node_leader => false
  },
  telegraf => {
    nodes => [
      "10.129.4.39:9273",
      "10.129.5.106:9273",
      "10.129.6.110:9273",
      "10.129.6.96:9273",
      "10.129.6.232:9273",
      "10.129.7.233:9273",
      "10.129.5.194:9273",
      "10.129.7.210:9273",
      "10.129.4.85:9273",
      "10.129.7.92:9273",
      "10.129.5.235:9273",
      "10.129.4.135:9273"
    ],
    nodes_string => "10.129.4.39:9273,10.129.5.106:9273,10.129.6.110:9273,10.129.6.96:9273,10.129.6.232:9273,10.129.7.233:9273,10.129.5.194:9273,10.129.7.210:9273,10.129.4.85:9273,10.129.7.92:9273,10.129.5.235:9273,10.129.4.135:9273",
    is_current_node_leader => false
  }
}
```

This allows us to set mongodb connection string as well as detect that current graylog node should be a leader based on kv lock.

The fact would return null in case of a failure to connect so should be safe for everyone but would indeed try to connect. Maybe there's some containment that could be added - let me know if you think of something.